Fix build with -DENABLE_LIBRARIES=0

rschu1ze opened the pull request:
- Replxx: When disabled via -DENABLE_LIBRARIES=0 or -DENABLE_REPLXX (the latter was undocumented) the build broke because replxx symbols were used since [0] in header LineReader.h. This header should in theory stay clean of replxx but doesn't for efficiency reasons. This change makes compilation of replxx mandatory. As replxx is quite small, I guess this is okay. (The alternative is to litter the code with ifdefs for non-replxx and a replxx paths.)

- OpenSSL: can no longer be disabled for now, too many 3rd party libs link it unconditionally, needs investigation

Also:
- removed per-contrib warning flags as warnings are disabled in contrib/CMakeLists.txt already
- removed optional M_LIBRARY linking which is obsolete
- added comments and cosmetics here and there
- Remove hint to checkout submodules/contribs recursively (verified locally that this works by building from a freshly cloned ClickHouse and "flat" checkout of submodules without recursion)

[0] https://github.com/ClickHouse/ClickHouse/pull/33201

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

rschu1ze committed:
Fix build with -DENABLE_LIBRARIES=0 / -DENABLE_REPLXX=0

Replxx: When disabled via -DENABLE_LIBRARIES=0 or -DENABLE_REPLXX (the
latter was undocumented) the build broke because replxx symbols were
used since [0] in header LineReader.h. This header should in theory
stay clean of replxx but doesn't for efficiency reasons.

This change makes compilation of replxx mandatory. As replxx is quite
small, I guess this is okay. (The alternative is to litter the code
with ifdefs for non-replxx and a replxx paths.)

[0] https://github.com/ClickHouse/ClickHouse/pull/33201

rschu1ze committed:
Cosmetics

rschu1ze committed:
Remove unused M_LIBRARY link

rschu1ze committed:
Removed warning flags in contribs

warnings are disabled for all contribs in contrib/CMakeLists.txt already

alexey-milovidov reviewed (commented):
Everything else LGTM.

rschu1ze commented on contrib/CMakeLists.txt:
```diff
@@ -38,7 +38,7 @@ function(add_contrib cmake_folder)
         file(GLOB contrib_files "${base_folder}/*")
         if (NOT contrib_files)
             message(STATUS "submodule ${base_folder} is missing or empty. to fix try run:")
-            message(STATUS "    git submodule update --init")
```
Ah, I remember you removed recursive submodules a while ago. Will fix then.

rschu1ze committed:
Make SSL a mandatory dependency for now

- SSL is a dependency of too many libs + unit tests (via poco crypto
  which requires SSL)

- optional SSL is desirable but right now, turning off SSL (via
  -DENABLE_LIBRARIES=0 or =DENABLE_SSL=0) breaks the build

- therefore make SSL mandatory for now + add a TODO comment

rschu1ze committed:
No need to checkout submodules/contribs recursively

Also verified locally by building from a freshly cloned ClickHouse and
"flat" checkout of submodules without recursion
